hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/GrpcBlockItemWriter.java

```diff
@@ -0,0 +1,214 @@
+    private static final String INVALID_MESSAGE = "Invalid protocol buffer converting %s from PBJ to protoc for %s";
+
+    private final BlockStreamServiceGrpc.BlockStreamServiceStub asyncStub;
+    private final ManagedChannel channel;
+    private StreamObserver<PublishStreamRequest> requestObserver;
+
+    private long blockNumber;
+
+    /** The state of this writer */
+    private State state = State.UNINITIALIZED;
+
+    /**
+     * The current state of the gRPC writer.
+     */
+    public enum State {
+        /**
+         * The gRPC client still not initialized
+         */
+        UNINITIALIZED,
+        /**
+         * The gRPC client is currently open
+         */
+        OPEN,
+        /**
+         * The gRPC client is already closed
+         */
+        CLOSED
+    }
+
+    /**
+     * @param blockStreamConfig the block stream configuration
+     */
+    public GrpcBlockItemWriter(@NonNull final BlockStreamConfig blockStreamConfig) {
+        requireNonNull(blockStreamConfig, "The supplied argument 'blockStreamConfig' cannot be null!");
+
+        channel = ManagedChannelBuilder.forAddress(blockStreamConfig.address(), blockStreamConfig.port())
+                .usePlaintext()
+                .build();
+        asyncStub = BlockStreamServiceGrpc.newStub(channel);
+    }
+
+    @Override
+    public void openBlock(long blockNumber) {
+        if (state != State.UNINITIALIZED) throw new IllegalStateException("GrpcBlockItemWriter initialized twice");
+
+        if (blockNumber < 0) throw new IllegalArgumentException("Block number must be non-negative");
+        this.blockNumber = blockNumber;
+        requestObserver = asyncStub.publishBlockStream(new StreamObserver<>() {
+            @Override
+            public void onNext(PublishStreamResponse streamResponse) {
+                // close the stream if error stream response occurs - source handshake design
+                // https://github.com/hashgraph/hedera-block-node/issues/120
+                /*if (streamResponse.getStatus().getStatus() == STREAM_ITEMS_UNKNOWN) {
+                 */
+                /*, STREAM_ITEMS_SUCCESS, STREAM_ITEMS_BEHIND*/
+                /*
+                    onNext(PublishStreamResponse.newBuilder()
+                            .setStatus(EndOfStream.newBuilder()
+                                    .setStatus(STREAM_ITEMS_UNKNOWN)
+                                    .build())
+                            .build());
+
+                    closeBlock();
+                }*/
+
+                if (streamResponse.hasAcknowledgement()) {
+                    final Acknowledgement acknowledgement = streamResponse.getAcknowledgement();
+                    if (acknowledgement.hasBlockAck()) {
+                        logger.info("PublishStreamResponse: a full block received: {}", acknowledgement.getBlockAck());
+                    } else if (acknowledgement.hasItemAck()) {
```

Comment:
Do we want to keep a series of booleans or something, representing that all block items have been acknowledged? Not sure if that's overkill or not here, just thinking aloud. Thoughts?

Reply:
Currently, the intention of those are to show that the connection between CN and BN is established and we can receive items from block nodes after they receive a stream. In the future there should be some handling mechanisms on different scenarios based on the handshake approach, which is yet to be finalized